client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,401 @@
-pub const CONSENSUS_TRACE: &str = "9";
-pub const CONSENSUS_DEBUG: &str = "5";
-pub const CONSENSUS_WARN: &str = "4";
-pub const CONSENSUS_INFO: &str = "1";
-
-/// Telemetry object. Implements `Future` and must be polled regularly.
-/// Contains an `Arc` and can be cloned and pass around. Only one clone needs to be polled
-/// regularly and should be polled regularly.
-/// Dropping all the clones unregisters the telemetry.
-#[derive(Clone)]
-pub struct Telemetry {
-	inner: Arc<Mutex<TelemetryInner>>,
-	/// Slog guard so that we don't get deregistered.
-	_guard: Arc<slog_scope::GlobalLoggerGuard>,
-}
-
-/// Behind the `Mutex` in `Telemetry`.
-///
-/// Note that ideally we wouldn't have to make the `Telemetry` cloneable, as that would remove the
-/// need for a `Mutex`. However there is currently a weird hack in place in `sc-service`
-/// where we extract the telemetry registration so that it continues running during the shutdown
-/// process.
-struct TelemetryInner {
-	/// Worker for the telemetry. `None` if it failed to initialize.
-	worker: Option<worker::TelemetryWorker>,
-	/// Receives log entries for them to be dispatched to the worker.
-	receiver: mpsc::Receiver<async_record::AsyncRecord>,
-}
-
-/// Implements `slog::Drain`.
-struct TelemetryDrain {
-	/// Sends log entries.
-	sender: std::panic::AssertUnwindSafe<mpsc::Sender<async_record::AsyncRecord>>,
-}
-
-/// Initializes the telemetry. See the crate root documentation for more information.
-///
-/// Please be careful to not call this function twice in the same program. The `slog` crate
-/// doesn't provide any way of knowing whether a global logger has already been registered.
-pub fn init_telemetry(config: TelemetryConfig) -> Telemetry {
-	// Build the list of telemetry endpoints.
-	let (endpoints, wasm_external_transport) = (config.endpoints.0, config.wasm_external_transport);
-
-	let (sender, receiver) = mpsc::channel(16);
-	let guard = {
-		let logger = TelemetryDrain { sender: std::panic::AssertUnwindSafe(sender) };
-		let root = slog::Logger::root(slog::Drain::fuse(logger), slog::o!());
-		slog_scope::set_global_logger(root)
-	};
-
-	let worker = match worker::TelemetryWorker::new(endpoints, wasm_external_transport) {
-		Ok(w) => Some(w),
-		Err(err) => {
-			error!(target: "telemetry", "Failed to initialize telemetry worker: {:?}", err);
-			None
+	/// Run the telemetry worker.
+	///
+	/// This should be run in a background task.
+	pub async fn run(self) {
+		let Self {
+			receiver,
+			sender: _sender,
+			mut init_receiver,
+			init_sender: _init_sender,
+			transport,
+		} = self;
+
+		let mut node_map: HashMap<Id, Vec<(u8, Multiaddr)>> = HashMap::new();
+		let mut connection_messages: HashMap<Multiaddr, Vec<ConnectionMessage>> = HashMap::new();
+		let mut connection_notifiers: HashMap<Multiaddr, Vec<ConnectionNotifierSender>> =
+			HashMap::new();
+		let mut existing_nodes: HashSet<Multiaddr> = HashSet::new();
+
+		// initialize the telemetry nodes
+		init_receiver.close();
```

Comment:
Why is that an unbounded channel? 

Why not just an oneshot channel?

Reply:
When you initialize multiple substrate nodes you will send multiple messages that will not be consumed until the telemetry actually starts. So it's unbounded (as you can init as many telemetries you want) but it's okay because it's really limited (one message per substrate node but also limited in time because it will be dropped early in the process execution)

Reply:
So that means if the worker starts earlier, because of whatever reason, some configuration will not be send? That is a bad idea :see_no_evil:

Reply:
The worker is started by the runner. It can only be at the end of your initialization XD How on earth would you make it start earlier?? https://github.com/paritytech/substrate/blob/cecton-telemetry-without-global/client/cli/src/runner.rs#L191

Reply:
I don't assume anything, it is impossible to run the telemetry worker without having initialized.

As I said: the telemetry worker is ran by the runner in sc-cli. It is 100% certain it will be the right order.

Reply:
You don't read what I have written.

I'm speaking about a child-node, like the polkadot node is for the parachain node. Your current setup requires that the initialization of the polkadot node is finished, before we finish the initialization of the parachain node. What I'm speaking about is that the polkadot node would be finish its initialization after the parachain node is initialized. In this case the polkadot telemetry would not be connected.

Reply:
I just checked the code of cumulus on master and it's true that the span won't work properly.

Last time I worked on cumulus things were different and it would have worked:

```rust
pub fn start_collator<'a, Block, PF, BI, BS, Client, Backend>(...) -> sc_service::error::Result<()> where ...
{
	let builder = CollatorBuilder::new(
		proposer_factory,
		inherent_data_providers,
		block_import,
		block_status,
		para_id,
		client,
		announce_block,
		block_announce_validator,
	);

	// <---- span would have started after this call
	let (polkadot_future, polkadot_task_manager) =
		polkadot_collator::start_collator(builder, para_id, collator_key, polkadot_config)?;

	task_manager
		.spawn_essential_handle()
		.spawn("polkadot", polkadot_future);

	task_manager.add_child(polkadot_task_manager);

	Ok(())
}
```

But now because the polkadot node is created early: https://github.com/paritytech/cumulus/blob/master/rococo-parachains/src/service.rs#L119

All the code after will be improperly spanned.

I'm not sure how I will fix this... will have to think more about it.

> What I'm speaking about is that the polkadot node would be finish its initialization after the parachain node is initialized. In this case the polkadot telemetry would not be connected.

For what I see both telemetries will be running and working because they are initialized in [this function](https://github.com/paritytech/cumulus/blob/master/rococo-parachains/src/service.rs#L98) and the telemetry worker is ran only after the closure of the run_node_until_exit [here](https://github.com/paritytech/cumulus/blob/master/rococo-parachains/src/command.rs#L269) finishes.

So I still don't see a problem with the initialization as you said. But there is a problem of span. But you do mean there is a problem of the polkadot node not being able to initialize because the telemetry worker is already started, right? I don't see this, maybe I'm missing something.

Reply:
To go around the issue with entering the span I separated it from the TelemetryHandle:
 - `TelemetryHandle` can only be used to initialize the telemetry (it doesn't enter the span anymore)
 - `TelemetryHandle` consumes itself when it initializes (avoid initializing multiple times for the same span)
 - There is a new type called `TelemetrySpan` which is just a `tracing::Span`. It is a handle and it can be cloned easily. With it we can re-enter the span easily. It is accessible from `Configuration`. (This solves any issue of span while initializing.)
 - Instead of entering the span in `spawn_tasks`, I made the `TaskManager` aware of the telemetry span and it will make all task spawned entered the span automatically.

These changes are in f6fa4a1c3

All of this answer the issue with the span but **does not** answer the issue you are mentioning with the initialization. I'm still not sure why you think the worker would run before things get initialized. I verified the code in cumulus, polkadot and substrate and I don't see how.

Reply:
Let's imagine the following.

1. Setup the node, get the handle and start the node. 
2. At some later point, maybe after some event we start a second node that uses the handle from 1.

As far as I understand your code this would not work, because the worker already finished its init phase?

Reply:
> maybe after some event

What? I didn't know we need to support that.

So you mean that a second node could start, triggered by an event during the first node running? (not sure if that's proper english)...

Ok no this is unsupported. Cumulus' nodes are started and stopped together.

So... I need to support that?